design-proposals/vm-rollout-strategy/vm-rollout-strategy.md

```diff
@@ -0,0 +1,195 @@
+# Overview
+
+Today changes to a VM object (i.e. changing VM.spec.template) get applied, but only take effect upon the next restart - we call these pending changes “staged”. This is a proposal to introduce a rollout strategy field to control how changes to the VM object will propagate to the VirtualMachineInstance object.
+
+
+## Motivation
+
+In most cases, the user would prefer the changes to the VM object to immediately propagate to the VMI object, as long as the rollout does not interrupt the workload. This behavior would be similar to how the Deployments roll out DaemonSets.
+
+
+## Goals
+
+-   Provide a way to propagate changes from a running VM object to its corresponding VMI immediately without interrupting the workload.
+    
+-   Maintain backwards compatibility
+    
+
+## Non Goals
+
+-   Successfully apply all the changes to the Virtual Machine Instance Object. Not all fields can be propagated to the guest immediately. Some fields should remain staged, such as those that affect the guest ABI (e.g. machine type)
+    
+-   Make LiveUpdate the default. We defer this decision.
+    
+
+## Definition of Users
+
+A user - a VM owner - with (at a minimum) create, update, and patch access for VM objects within a namespace.
+
+
+## User Stories
+
+As a user I want changes that I applied to the VM object to propagate to the VMI immediately if possible.
+
+As a user, I want to be notified about changes that require a restart of the VM.
+
+## Repos
+
+kubevirt/kubevirt
+
+
+# Design
+
+A new field called rolloutStrategy will be introduced in the VM spec. This field will control the behavior of the VM object. Two possible values will be available:
```

Comment:
Can we add a section that is explaining why a new VM API field is the only way out?
Was it considered to start with cluster level (KubeVirt CR level) field only, and not extend the VM API?

If it was discussed, please decsribe why. If it was not discussed, then ... I'd encourage us to do so :)

Reply:
A thought: it makes sense to me that some workloads must avoid migrations unless necessary. While I think that staging updates to the VM is probably not that useful in general, I think that using `stage` strategy is especially valuable to avoid being surprised by a migration.

Saying that, EdDev's idea regarding using selectors sounds reasonable and I think we should consider it.